### PR TITLE
Problem: hazardous cloud backup/recovery option (fixes #79)

### LINKED ISSUE #79: Problem: hazardous cloud backup/recovery option
currently, one needs a symmetric key in plaintext on the machine temporarily

### PR DESCRIPTION
**Pull request by tomtau**
Solution:
- added an option to generate an RSA key pair for wrapping that should be compatible with potential cloud provider's services
- historical key unwrapping using RFC3394/RFC5649
- currently, `rsa` crate requires a new version of `rand` which is in conflict with the current `ed25519-dalek`, hence the patched version
- changed init/recover functionality to use the asymmetric option

### CODE REVIEW

File: providers/sgx/sgx-app/src/sgx_app/cloud.rs [hunk 1]
```diff
@@ -0,0 +1,352 @@
...
+
+        let rdata = hasher.finalize();
+        let x: &[u8] = &rdata;
+        let y: &[u8] = &report.reportdata[..32];
+        assert_eq!(x, y);
+    }
+
+    #[test]
+    fn test_rsa2() {
+        let file_content = r#"
```
Comment (tomtau) [inline providers/sgx/sgx-app/src/sgx_app/cloud.rs:268]: should add a comment it's a test-only key :)

File: providers/sgx/sgx-app/src/sgx_app/cloud.rs [hunk 2]
```diff
@@ -0,0 +1,352 @@
...
+
+    let claim_payload = get_claim(&pub_key);
+    let mut hasher = Sha256::new();
+    hasher.update(claim_payload);
+    let result = hasher.finalize();
+    let result_len = result.as_slice().len();
+    let mut report_data = [0u8; 64];
+    for i in 0..result_len {
+        report_data[i] = result[i];
+    }
```
[the remark this answers is not in the document]
Comment (tomtau) [reply providers/sgx/sgx-app/src/sgx_app/cloud.rs:59]: it's the output of sha256, so it should always be -- the result's should be `GenericArray<u8, U32>` -- I can put an explicit annotation there as a sanity check / compile-time check

File: providers/sgx/sgx-app/src/sgx_app/cloud.rs [hunk 3]
```diff
@@ -0,0 +1,352 @@
...
+}
+
+/// Generates an RSA keypair (2048bit with e=65537),
+/// seals the private key to a local CPU,
+/// and gets a report for sha256 of the public key in JWE.
+pub fn generate_keypair(
+    csprng: &mut OsRng,
+    targetinfo: Targetinfo,
+) -> Result<(RSAPublicKey, SealedKeyData, Report), CloudError> {
+    let bits = 2048;
```
[the remark this answers is not in the document]
Comment (tomtau) [reply providers/sgx/sgx-app/src/sgx_app/cloud.rs:45]: 👌

File: providers/sgx/sgx-app/src/sgx_app/cloud.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (linfeng-crypto) [reply providers/sgx/sgx-app/src/sgx_app/cloud.rs:59]: so we can use like this: `let report_data = result.try_into.unwrap();`?

[the remark this answers is not in the document]
Comment (tomtau) [reply providers/sgx/sgx-app/src/sgx_app/cloud.rs:59]: yeah, maybe we can re-write it as `report_data[0..32].copy_from_slice(&result)`

**Review by allthatjazzleo — APPROVED** [review]

File: providers/sgx/sgx-app/src/sgx_app/cloud/keywrap.rs [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+    let kek = GenericArray::from_slice(kek);
+    let wrapped = GenericArray::from_slice(wrapped);
+    let (mut key, key_iv) = aes256_unwrap_key_and_iv(kek, wrapped);
+
+    // Alternate initial value for aes key wrapping, as defined in RFC 5649 section 3
+    // http://www.ietf.org/rfc/rfc5649.txt
+    // big-ending length hardcoded to 32 (as here, it's always unwrapping 32-byte sealing key)
+    const IV_5649: [u8; 8] = [0xa6, 0x59, 0x59, 0xa6, 0, 0, 0, 32];
+
+    if IV_5649.ct_eq(&key_iv).unwrap_u8() == 0u8 {
```
Comment (durgeshp-crypto) [inline providers/sgx/sgx-app/src/sgx_app/cloud/keywrap.rs:84]: +1

**Review by durgeshp-crypto — COMMENTED** [review]
Changes in keywrap.rs looks good to me and aligned to rfc3394.txt section 2.2.2

**Pull request merged**